docs/kf_uids/uid_considerations.md

```diff
@@ -0,0 +1,220 @@
+* **Randomness never guarantees uniqueness.** Ever. Regardless of the number space or generator strength. We can do better than that.
+* None of the 5 basic UUID/GUID variants described on wikipedia.org/wiki/Universally_unique_identifier ensure uniqueness. They are all misnamed.
+
+## We definitely want
+
+* Eliminate key reuse and collisions
+* Compact representation
+* Visually differentiable (no commonly hyperambiguous characters)
+* Typable
+
+## We don't need
+
+(many systems value these things for sometimes legitimate and sometimes illegitimate reasons)
+
+* Chronological sorting (though recording time is useful for uniqueness)
+* Reading out loud over the phone
+* Easy typing on mobile keyboards
+* Checksums
+
+## We don't know if we need
+
+* Decentralized generation (could rule out keyservers)
+
+---
+
+## Key Generation Techniques
+
+(randomness, external incrementer, internal incrementer, physical ID, logical ID)
+
+### Methods for differentiating my own keys from each other
+
+* Random generation in a very large space
+  * <font color="darkorange">Insufficient. Improbable things happen all the time.</font>
+* Counting
+  * <font color="lightseagreen">Guaranteed for serial generation within max_increment.</font>
+* Timestamp
+  * <font color="lightseagreen">Guaranteed for extra-interval generation.</font>
+
+### Methods for differentiating my keys from someone else's keys
+
+* Random generation in a very large space
+  * <font color="darkorange">Insufficient. Improbable things happen all the time.</font>
+* Generator hardware address (e.g. prefix with MAC)
+  * <font color="crimson">Hardware address may not be stable. Hardware address is also the same for all processes on one device. Both defeat the purpose.</font>
+* Unique process ID per generator
+  * <font color="lightseagreen">Guaranteed, but requires negotiation to assign a unique ID to each process.</font>
+
+### Who uses what? ULID vs TUID vs Twitter vs Instagram vs Flickr, etc
+
+* **Twitter (snowflake)** uses <font color="lightseagreen">a timestamp, a node ID, and a counter</font>
+* **Instagram** uses a <font color="lightseagreen">timestamp, a node ID, and a counter</font>
+* **Flickr** uses <font color="lightseagreen">a central ticket server</font> with a simple <font color="lightseagreen">counter</font>
+  * (technically multiple servers where the number space is subdivided among them to avoid single-point-of-failure)
+* **TUID** uses <font color="lightseagreen">a timestamp, a node ID, a counter, <font color="darkorange">and some random bits</font></font>
+* **ULID** uses <font color="lightseagreen">a timestamp and a counter <font color="darkorange">that starts randomized</font></font>
+* **MongoDB ObjectId** uses <font color="lightseagreen">a timestamp, <font color="darkorange">some random bits</font>, and a counter <font color="darkorange">that starts randomized</font></font>
+
+### Node IDs
```

Comment:
Is a node a process in this case?

Reply:
Or a device (if a given device only runs one generating process at a time) or a session.